File: algorithms/2_medium/M/04008_minimum_initial_strength_to_defeat_all_monsters/prefix.c
long long minInitialStrength(
    const int* monsters, const int monsters_len,
    int** boosts, const int boosts_len, const int* boosts_cols_data
) {
    if (monsters == NULL || monsters_len <= 0) {
        fprintf(stderr, "invalid input");
        goto err;
    }

    if (boosts_len > 0 && (boosts == NULL || boosts_cols_data == NULL)) {
        fprintf(stderr, "invalid input");
        goto err;
    }

    int64_t* prefix = calloc(((size_t)monsters_len + 1), sizeof(*prefix));
    if (prefix == NULL) {
        fprintf(stderr, "alloc fail");
        goto err;
    }

    for (int i = 0; i < boosts_len; i += 1) {
        if (boosts_cols_data[i] != 3) {
            fprintf(stderr, "invalid input");
            goto cleanup_err;
        }
        const int* b = boosts[i];
        const int l = b[0];
        const int r = b[1];
        const int v = b[2];
        if (l < 0 || r < l || v < 0 || r >= monsters_len) {
            fprintf(stderr, "invalid input");
            goto cleanup_err;
        }
        prefix[l] += v;
        prefix[r + 1] -= v;
    }

    for (int i = 1; i < monsters_len; i += 1) {
        prefix[i] += prefix[i - 1];
    }

    int idx = monsters_len - 1;
    while (idx >= 0 && prefix[idx] >= monsters[idx]) {
        idx -= 1;
    }

    int64_t curr = 0;
    if (idx >= 0) {
        curr += monsters[idx] - prefix[idx];
        idx -= 1;
    }
    while (idx >= 0) {
        curr += monsters[idx];
        idx -= 1;
    }

/* cleanup_ok: */
    free(prefix);
    return curr;

cleanup_err:
    free(prefix);
err:
    exit(1);
}

File: algorithms/2_medium/M/04008_minimum_initial_strength_to_defeat_all_monsters/prefix.rs
impl Solution {
    pub fn min_initial_strength(monsters: Vec<i32>, boosts: Vec<Vec<i32>>) -> i64 {
        let monsters_len = monsters.len();
        let mut prefix = vec![0_i64; monsters_len + 1];
        for b in boosts {
            assert!(b.len() == 3);
            let (l, r, v) = (b[0], b[1], b[2]);
            assert!(l >= 0 && r >= l && v >= 0);
            let (l, r) = (l as usize, r as usize);
            assert!(r < monsters_len);
            prefix[l] += v as i64;
            prefix[r + 1] -= v as i64;
        }
        prefix.pop();
        for i in 1..monsters_len {
            prefix[i] += prefix[i - 1];
        }

        let mut idx = monsters_len as isize - 1;
        while idx >= 0 && prefix[idx as usize] >= monsters[idx as usize] as i64 {
            idx -= 1;
        }

        let mut curr = 0_i64;
        if idx >= 0 {
            curr += monsters[idx as usize] as i64 - prefix[idx as usize];
            idx -= 1;
        }
        while idx >= 0 {
            curr += monsters[idx as usize] as i64;
            idx -= 1;
        }
        curr
    }
}
